From 2e4c268b1cc124e8cf3b0782a3bd8a4d19343daf Mon Sep 17 00:00:00 2001
From: Eliad Peller <eliad@wizery.com>
Date: Thu, 19 Dec 2013 13:22:10 +0200
Subject: [PATCH 07/47] hostapd: sync channel on interface setup [AP AP]

If another ap iface is enabled, sync to its channel
(if ap_channel_sync is set).

Signed-off-by: Eliad Peller <eliad@wizery.com>
---
 src/ap/hostapd.c | 33 +++++++++++++++++++++++++++++++++
 1 file changed, 33 insertions(+)

diff --git a/src/ap/hostapd.c b/src/ap/hostapd.c
index bf1975f..9ce1bf1 100644
--- a/src/ap/hostapd.c
+++ b/src/ap/hostapd.c
@@ -1624,6 +1624,35 @@ fail:
 	return -1;
 }
 
+static int hostapd_sync_channel(struct hostapd_iface *hapd_iface)
+{
+	int i;
+
+	if (!hapd_iface->interfaces)
+		return -1;
+
+	wpa_printf(MSG_DEBUG, "Syncing AP channel");
+
+	for (i = 0; i < hapd_iface->interfaces->count; i++) {
+		struct hostapd_iface *iface = hapd_iface->interfaces->iface[i];
+
+		if (iface->state != HAPD_IFACE_ENABLED)
+			continue;
+
+		hapd_iface->conf->hw_mode = iface->conf->hw_mode;
+		hapd_iface->conf->channel = iface->conf->channel;
+		hapd_iface->conf->secondary_channel =
+				iface->conf->secondary_channel;
+		wpa_printf(MSG_DEBUG, "Channel automatically synced to "
+			   "existing AP: %d (secondary: %d) (mode: %s)",
+			   iface->conf->channel,
+			   iface->conf->secondary_channel,
+			   hostapd_hw_mode_txt(iface->conf->hw_mode));
+		return 1;
+	}
+
+	return 0;
+}
 
 #ifdef CONFIG_FST
 
@@ -1873,6 +1902,10 @@ static int hostapd_setup_interface_complete_sync(struct hostapd_iface *iface,
 		goto fail;
 
 	wpa_printf(MSG_DEBUG, "Completing interface initialization");
+
+	if (iface->conf->ap_channel_sync)
+		hostapd_sync_channel(iface);
+
 	if (iface->conf->channel) {
 #ifdef NEED_AP_MLME
 		int res;
-- 
1.9.1

